### ISSUE TITLE
capnproto 0.10.3 fails to compile with clang-16 and C++20

### ISSUE THREAD

**Issue by Hixos** [capnproto/capnproto#1622]
I'm trying to build the latest release (0.10.3) on Linux with clang-16 and C++20 (passing `-DCMAKE_CXX_STANDARD=20` to CMake), and I get the following error:

```
/home/luca/code/capnproto-c++-0.10.3/src/capnp/blob.h:177:74: error: invalid operands to binary expression ('const char *' and 'kj::StringPtr')
inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
                                                                       ~ ^  ~~~~~~~~~~~~
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:756:13: note: candidate function template not viable: no known conversion from 'kj::StringPtr' to 'MaxValue_' for 2nd argument
inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
            ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:756:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char *' to 'MaxValue_' for 1st argument
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:758:13: note: candidate function template not viable: no known conversion from 'kj::StringPtr' to 'MinValue_' for 2nd argument
inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
            ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:758:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char *' to 'MinValue_' for 1st argument
/home/luca/code/capnproto-c++-0.10.3/src/capnp/blob.h:176:13: note: candidate function not viable: no known conversion from 'kj::StringPtr' to 'const Text::Builder' for 2nd argument
inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
            ^
/home/luca/code/capnproto-c++-0.10.3/src/capnp/blob.h:177:13: note: candidate function not viable: no known conversion from 'kj::StringPtr' to 'const Text::Builder' for 2nd argument
inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
            ^                                                               ~~~~~~~~~~~~
```

I tried with clang-17 as well and the same issue arises, while clang-10 or GCC 11 seem to work fine.
  
I was able to temporarily work around this issue when linking capnproto with my project by manually defining the `bool operator==(const char* other)` function in the `kj::StringPtr` and `kj::String` classes, but the problem is not completely solved when compiling capnproto directly as I now get errors in `debug.h`:  
  
```
/home/luca/code/capnproto-c++-0.10.3/src/kj/debug.h:654:19: error: invalid operands to binary expression ('kj::StringPtr' and 'kj::ArrayPtr<const char>')
  DEFINE_OPERATOR(==);
  ~~~~~~~~~~~~~~~~^~~
/home/luca/code/capnproto-c++-0.10.3/src/kj/debug.h:651:25: note: expanded from macro 'DEFINE_OPERATOR'
    bool result = value OP other; \
                  ~~~~~ ^  ~~~~~
/home/luca/code/capnproto-c++-0.10.3/src/kj/string-test.c++:250:15: note: in instantiation of function template specialization 'kj::_::DebugExpression<kj::StringPtr &>::operator==<kj::ArrayPtr<const char> &>' requested here
  KJ_EXPECT(s == a);
              ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:756:13: note: candidate function template not viable: no known conversion from 'kj::ArrayPtr<const char>' to 'MaxValue_' for 2nd argument
inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
            ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:756:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'kj::StringPtr' to 'MaxValue_' for 1st argument
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:758:13: note: candidate function template not viable: no known conversion from 'kj::ArrayPtr<const char>' to 'MinValue_' for 2nd argument
inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
            ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/common.h:758:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'kj::StringPtr' to 'MinValue_' for 1st argument
/home/luca/code/capnproto-c++-0.10.3/src/kj/string.h:124:25: note: candidate function not viable: no known conversion from 'kj::ArrayPtr<const char>' to 'decltype(nullptr)' (aka 'std::nullptr_t') for 1st argument
  inline constexpr bool operator==(decltype(nullptr)) const { return content.size() <= 1; }
                        ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/string.h:563:24: note: candidate function not viable: no known conversion from 'kj::ArrayPtr<const char>' to 'const StringPtr' for 1st argument
inline bool StringPtr::operator==(const StringPtr& other) const {
                       ^
/home/luca/code/capnproto-c++-0.10.3/src/kj/string.h:129:15: note: candidate function not viable: no known conversion from 'kj::ArrayPtr<const char>' to 'const char *' for 1st argument
  inline bool operator==(const char* other) const
              ^
```

**Comment by kentonv** [comment]
Does #1618 fix this?

**Comment by Hixos** [comment]
Partially: it solves the first compilation error and I'm able to compile the library, but tests still fail with many instances of the second error.

**Comment by pavel-odintsov** [comment]
I can confirm that exactly same issue happens on Fedora 38 with gcc 13:
```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/13/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,m2,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --enable-libstdcxx-backtrace --with-libstdcxx-zoneinfo=/usr/share/zoneinfo --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-13.0.1-20230215/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-offload-defaulted --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux --with-build-config=bootstrap-lto --enable-link-serialization=1
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.0.1 20230215 (Red Hat 13.0.1-0) (GCC)
```

gcc's error message looks little bit differnt way:
```
In file included from /usr/include/capnp/layout.h:34,
                 from /usr/include/capnp/generated-header-support.h:27,
                 from /home/odintsov/rpmbuild/BUILD/fastnetmon-master/src/simple_packet_capnp/simple_packet.capnp.h:6,
                 from /home/odintsov/rpmbuild/BUILD/fastnetmon-master/src/simple_packet_capnp/simple_packet.capnp.c++:4:
/usr/include/capnp/blob.h: In function 'bool capnp::operator==(const char*, const Text::Builder&)':
/usr/include/capnp/blob.h:176:74: error: no match for 'operator==' (operand types are 'const char*' and 'kj::StringPtr')
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |                                                                        ~ ^~ ~~~~~~~~~~~~
      |                                                                        |              |
      |                                                                        const char*    kj::StringPtr
In file included from /usr/include/kj/memory.h:24,
                 from /usr/include/kj/array.h:24,
                 from /usr/include/kj/string.h:25,
                 from /usr/include/capnp/common.h:29,
                 from /usr/include/capnp/raw-schema.h:24,
                 from /usr/include/capnp/generated-header-support.h:26:
/usr/include/kj/common.h:756:13: note: candidate: 'template<class T> bool kj::operator==(T, MaxValue_)' (reversed)
  756 | inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
      |             ^~~~~~~~
/usr/include/kj/common.h:756:13: note:   template argument deduction/substitution failed:
/usr/include/capnp/blob.h:176:72: note:   cannot convert 'a' (type 'const char*') to type 'kj::MaxValue_'
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |                                                                        ^
/usr/include/kj/common.h:758:13: note: candidate: 'template<class T> bool kj::operator==(T, MinValue_)' (reversed)
  758 | inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
      |             ^~~~~~~~
/usr/include/kj/common.h:758:13: note:   template argument deduction/substitution failed:
/usr/include/capnp/blob.h:176:72: note:   cannot convert 'a' (type 'const char*') to type 'kj::MinValue_'
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |                                                                        ^
/usr/include/capnp/blob.h:176:13: note: candidate: 'bool capnp::operator==(const char*, const Text::Builder&)' (reversed)
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |             ^~~~~~~~
/usr/include/capnp/blob.h:176:36: note:   no known conversion for argument 1 from 'kj::StringPtr' to 'const char*'
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |                        ~~~~~~~~~~~~^
/usr/include/capnp/blob.h: In function 'bool capnp::operator!=(const char*, const Text::Builder&)':
/usr/include/capnp/blob.h:177:74: error: no match for 'operator!=' (operand types are 'const char*' and 'kj::StringPtr')
  177 | inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
      |                                                                        ~ ^~ ~~~~~~~~~~~~
      |                                                                        |              |
      |                                                                        const char*    kj::StringPtr
/usr/include/kj/common.h:756:13: note: candidate: 'template<class T> bool kj::operator==(T, MaxValue_)' (reversed)
  756 | inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
      |             ^~~~~~~~
/usr/include/kj/common.h:756:13: note:   template argument deduction/substitution failed:
/usr/include/capnp/blob.h:177:72: note:   cannot convert 'a' (type 'const char*') to type 'kj::MaxValue_'
  177 | inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
      |                                                                        ^
/usr/include/kj/common.h:758:13: note: candidate: 'template<class T> bool kj::operator==(T, MinValue_)' (reversed)
  758 | inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
      |             ^~~~~~~~
/usr/include/kj/common.h:758:13: note:   template argument deduction/substitution failed:
/usr/include/capnp/blob.h:177:72: note:   cannot convert 'a' (type 'const char*') to type 'kj::MinValue_'
  177 | inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
      |                                                                        ^
/usr/include/capnp/blob.h:176:13: note: candidate: 'bool capnp::operator==(const char*, const Text::Builder&)' (reversed)
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |             ^~~~~~~~
/usr/include/capnp/blob.h:176:36: note:   no known conversion for argument 1 from 'kj::StringPtr' to 'const char*'
  176 | inline bool operator==(const char* a, const Text::Builder& b) { return a == b.asString(); }
      |                        ~~~~~~~~~~~~^
/usr/include/capnp/blob.h:177:13: note: candidate: 'bool capnp::operator!=(const char*, const Text::Builder&)'
  177 | inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
      |             ^~~~~~~~
/usr/include/capnp/blob.h:177:60: note:   no known conversion for argument 2 from 'kj::StringPtr' to 'const capnp::Text::Builder&'
  177 | inline bool operator!=(const char* a, const Text::Builder& b) { return a != b.asString(); }
      |                                       ~~~~~~~~~~~~~~~~~~~~~^
```

Your proposed patch definitely addresses it. 

I manually applied your patch on 0.10.3-1 this way:
```
diff -u string.original string.h 
--- string.original	2023-02-21 17:35:30.424977256 +0000
+++ string.h	2023-02-21 17:38:17.000687845 +0000
@@ -122,10 +122,15 @@
   inline constexpr const char* end() const { return content.end() - 1; }
 
   inline constexpr bool operator==(decltype(nullptr)) const { return content.size() <= 1; }
+#if !__cpp_impl_three_way_comparison
   inline constexpr bool operator!=(decltype(nullptr)) const { return content.size() > 1; }
+#endif
 
   inline bool operator==(const StringPtr& other) const;
+#if !__cpp_impl_three_way_comparison
   inline bool operator!=(const StringPtr& other) const { return !(*this == other); }
+#endif 
+
   inline bool operator< (const StringPtr& other) const;
   inline bool operator> (const StringPtr& other) const { return other < *this; }
   inline bool operator<=(const StringPtr& other) const { return !(other < *this); }
```

And after these changes capnpn dependent tool compiled just fine. 

Thank you so much for your hard work in maintaining and creating this awesome library.

**Comment by kentonv** [comment]
Hello,

I have put together a release candidate with what I believe are all the known compiler fixes that have been merged:

http://capnproto.org/capnproto-c++-0.10.4-rc1.tar.gz

However, it's not clear to me from the comments here if there is still a problem or not. My distro doesn't have a readily-available gcc-13 package so it's a bit hard for me to test this. Could someone tell me if the above package works? If not, can someone send a PR for whatever is still needed here?

**Comment by Hixos** [comment]
Tried with the release candidate and clang-16, I can compile the library but still get the error when compiling the tests.  
Unfortunately I don't have GCC-13 available to try as well.

**Comment by kentonv** [comment]
I managed to get clang 16 installed locally and I cannot reproduce any problems compiling and running the tests with this release candidate. I'm going to go ahead and release this. If you could figure out what's different about your environment and send a PR I'll be happy to accept it and do another point release.

**Comment by topazus** [comment]
I failed to build capnproto 0.10.4 with clang 16 and set option `-DCMAKE_CXX_STANDARD=20` in cmake. detailed build log: https://kojipkgs.fedoraproject.org//work/tasks/5203/101655203/build.log

```bash
In file included from /builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common-test.c++:23:
In file included from /builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/test.h:24:
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/debug.h:654:19: error: invalid operands to binary expression ('const char[4]' and 'kj::String')
  DEFINE_OPERATOR(==);
  ~~~~~~~~~~~~~~~~^~~
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/debug.h:651:25: note: expanded from macro 'DEFINE_OPERATOR'
    bool result = value OP other; \
                  ~~~~~ ^  ~~~~~
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common-test.c++:334:5: note: in instantiation of function template specialization 'kj::_::DebugExpression<const char (&)[4]>::operator==<kj::String &>' requested here
    EXPECT_EQ("foo", s);
    ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/compat/gtest.h:56:39: note: expanded from macro 'EXPECT_EQ'
#define EXPECT_EQ(x, y) KJ_EXPECT((x) == (y), x, y)
                                      ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:756:13: note: candidate function template not viable: no known conversion from 'kj::String' to 'MaxValue_' for 2nd argument
inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
            ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:756:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char[4]' to 'MaxValue_' for 1st argument
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:758:13: note: candidate function template not viable: no known conversion from 'kj::String' to 'MinValue_' for 2nd argument
inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
            ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:758:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char[4]' to 'MinValue_' for 1st argument
In file included from /builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common-test.c++:23:
In file included from /builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/test.h:24:
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/debug.h:654:19: error: invalid operands to binary expression ('const char[4]' and 'kj::String')
  DEFINE_OPERATOR(==);
  ~~~~~~~~~~~~~~~~^~~
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/debug.h:651:25: note: expanded from macro 'DEFINE_OPERATOR'
    bool result = value OP other; \
                  ~~~~~ ^  ~~~~~
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common-test.c++:335:5: note: in instantiation of function template specialization 'kj::_::DebugExpression<const char (&)[4]>::operator==<kj::String>' requested here
    EXPECT_EQ("foo", kj::mv(m).orDefault([] {
    ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/compat/gtest.h:56:39: note: expanded from macro 'EXPECT_EQ'
#define EXPECT_EQ(x, y) KJ_EXPECT((x) == (y), x, y)
                                      ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:756:13: note: candidate function template not viable: no known conversion from 'kj::String' to 'MaxValue_' for 2nd argument
inline bool operator==(T t, MaxValue_) { return t == Decay<T>(maxValue); }
            ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:756:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char[4]' to 'MaxValue_' for 1st argument
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:758:13: note: candidate function template not viable: no known conversion from 'kj::String' to 'MinValue_' for 2nd argument
inline bool operator==(T t, MinValue_) { return t == Decay<T>(minValue); }
            ^
/builddir/build/BUILD/capnproto-0.10.4/c++/src/kj/common.h:758:13: note: candidate function (with reversed parameter order) template not viable: no known conversion from 'const char[4]' to 'MinValue_' for 1st argument
2 errors generated.
gmake[2]: *** [c++/src/kj/CMakeFiles/kj-tests.dir/build.make:79: c++/src/kj/CMakeFiles/kj-tests.dir/common-test.c++.o] Error 1
gmake[2]: Leaving directory '/builddir/build/BUILD/capnproto-0.10.4/redhat-linux-build'
gmake[1]: *** [CMakeFiles/Makefile2:1131: c++/src/kj/CMakeFiles/kj-tests.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

**Comment by kentonv** [comment]
Sorry but an error log doesn't help me fix the issue. To fix this I need someone to either:

a) Submit a PR which fixes the issue.
b) Submit a PR which modifies CI so that it reproduces the issue, so that I can then fix it.
